README.rst
.. contents::

Introduction
============

See the doc of http://pypi.python.org/pypi/five.z2monitor


Use zc.monitor and additional plugins to fetch probes via another thread than the one defined in Zope.

Once the instance is running zc.monitor thread listen to another port (127.0.0.1:8888 in this buildout). You can query values using simple python script or nc.

Example::

    echo 'uptime' | nc -i 1 localhost 8888


Or (when instance is up)::

    bin/instance monitor stats

Or::

    telnet 127.0.0.1 8888

    dbsize



Probes
======

Currently supported probes:

- cache_size -- cache sizes informations
- check_smtp -- Check if SMTP is initialize, return number of errors found.
- check_upgrade_steps -- Check if all upgrade steps are ran.
- conflictcount -- number of all conflict errors since startup
- count_users -- the total amount of users in your plone site
- count_valid_users -- Count all users connected since 90 days
- cpu_times -- ?
- creation_date_plonesite -- Get creation date of plonesite object. Default return unix_time (defaut=True) if you want ISO time call 'False' attr.
- dates -- Return all date probes
- dbactivity -- number of load, store and connections on database (default=main) for the last x minutes (default=5)
- dbinfo -- Get database statistics
- dbsize -- size of the database (default=main) in bytes
- errorcount -- number of error present in error_log (default in the root).
- health_ok -- fast health check for load balancers that simply returns 'OK'
- health_db_connected  -- returns the string 'OK' if database (default=main) is connected.
- help -- Get help about server commands
- interactive -- Turn on monitor's interactive mode
- last_login_time -- Get last login time user. Default return unix_time (defaut=True) if you want ISO time call 'False' attr.
- last_modified_plone_object_time -- Get last modified plone object time. Default return unix_time (defaut=True) if you want ISO time call 'False' attr.
- last_modified_zope_object_time -- Get last modified zope object time. Default return unix_time (defaut=True) if you want ISO time call 'False' attr.
- logstats -- ?
- memory_percent -- ?
- monitor -- Get general process info
- objectcount -- number of object in the database (default=main)
- quit -- Quit the monitor
- refcount -- the total amount of object reference counts
- requestqueue_size -- number of requests waiting in the queue to be handled by zope threads
- stats -- Stats of all information Products.ZNagios know
- threads -- Dump current threads execution stack
- unresolved_conflictcount -- number of all unresolved conflict errors since startup
- uptime -- uptime of the zope instance in seconds
- zeocache -- Get ZEO client cache statistics
- zeostatus -- Get ZEO client status information

How it works
============

This package use differents package

- five.z2monitor
- Products.ZNagios
- munin.zope
- zc.z3monitor
- zc.monitorcache
- zc.monitorlogstats

Add lines on your buildout::

    [instance]
    ...
    zope-conf-additional =
      <product-config five.z2monitor>
        bind 127.0.0.1:8888
      </product-config>



CHANGES.rst
Changelog
=========

0.2.7 (unreleased)
------------------

- Nothing changed yet.


0.2.6.1 (2018-02-01)
--------------------

- Close database connection after health check and improve haproxy documentation
  [fRiSi]


0.2.6 (2018-01-30)
------------------

- Add probes for load balancer health-checks (see doc/haproxy.rst for an example)
  [fRiSi]


0.2.5 (2016-02-25)
------------------

- Remove old threadframe dependency package.
  [bsuttor]


0.2.4 (2015-09-03)
------------------

- Add eggs probe.
  [bsuttor]


0.2.3 (2015-08-25)
------------------

- Return unix_time by default, you can still get ISO time with call parameters False.
  [bsuttor]


0.2.2 (2015-08-18)
------------------

- Fix error if last_login is empty.
  [bsuttor]


0.2.1 (2015-08-13)
------------------

- Return date in ISO format
  [bsuttor]


0.2.0 (2015-08-11)
------------------

- Add a probe (dates) which return all probes date.
  [bsuttor]

- Add last_login_time porbe.
  [bsuttor]

- Add last_modified_plone_object_time probe.
  [bsuttor]

- Add last_modified_zope_object_time probe.
  [bsuttor]


0.1.2 (2015-07-24)
------------------

- Date probe are not more in stats probe (from Products.ZNagios).
  [bsuttor]

- Add creation_date_plonesite probe
  [bsuttor]


0.1.1 (2015-07-22)
------------------

- Return number of smtp errors found.
  [bsuttor]


0.1 (2015-07-17)
----------------

- Add count_users probe
  [bsuttor]

- Initial package
  [bsuttor]


collective/monitor/monitor.py
# -*- coding: utf-8 -*-
from cStringIO import StringIO
from DateTime import DateTime
from Products.MailHost.interfaces import IMailHost
from Products.ZNagios.zcmonitor import beautify_return_values
from zc.z3monitor.interfaces import IZ3MonitorPlugin
from zope.component import getUtility, getUtilitiesFor
from zope.component.hooks import setSite
from zope.component.interfaces import ComponentLookupError
from Zope2 import app as App
import inspect
import os


def get_plone_site(connection, plone_path=None):
    """ Return plone site """
    container = App()
    # app = container.unrestrictedTraverse('/')
    if plone_path:
        plone_paths = [path for path in plone_path.split(os.sep) if path]
        for path_name in plone_paths:
            container = container.get(path_name)
        plone_site = container
        if not plone_site:
            msg = "Error, path '{0}' do not exist".format(plone_path)
            connection.write(str(msg))
            container._p_jar.close()
            return False
        if plone_site.meta_type != "Plone Site":
            msg = "Error, path {0} is not a plone site, it's {1}".format(plone_path, plone_site.meta_type)
            connection.write(str(msg))
            container._p_jar.close()
            return False
        else:
            return container, plone_site
    else:
        result = False
        for obj in container.values():
            if obj.meta_type == "Folder":  # if plonesite come from mount point
                for plone in obj.values():
                    if plone.meta_type == "Plone Site":
                        result = plone
            else:
                if obj.meta_type == "Plone Site" and not result:
                    result = obj
        return container, result


def get_users(context, obj=True):
    from Products.CMFCore.utils import getToolByName
    portal = getToolByName(context, "portal_url").getPortalObject()
    users = []
    for user in portal.acl_users.searchUsers():
        if user['pluginid'] == 'source_users':
            if obj:
                users.append(portal.portal_membership.getMemberById(user['userid']))
            else:
                users.append(user['userid'])
    return users


def count_users(connection, plone_path=None):
    """the total amount of users in your plone site"""
    app, plone_site = get_plone_site(connection, plone_path)
    if plone_site:
        setSite(plone_site)
        users = get_users(plone_site)
        connection.write(str(len(users)))
        app._p_jar.close()


def count_valid_users(connection, plone_path=None):
    """Count all users connected since 90 days"""
    app, plone_site = get_plone_site(connection, plone_path)
    if plone_site:
        setSite(plone_site)
        users = get_users(plone_site)
        valid_users = []
        for user in users:
            if user.getProperty("last_login_time") > (DateTime() - 90):
                valid_users.append(user)
        connection.write(str(len(valid_users)))
        app._p_jar.close()


def check_smtp(connection, plone_path=None):
    """Check if SMTP is initialize, return number of errors found. """
    app, plone_site = get_plone_site(connection, plone_path)
    if plone_site:
        setSite(plone_site)
        try:
            mail_host = getUtility(IMailHost)
        except ComponentLookupError:
            mail_host = getattr(plone_site, "MailHost")
        mail_errors = []
        if not mail_host.smtp_host:
            mail_errors.append("incorrect host: {0}".format(mail_host.smtp_host))
        if mail_host.smtp_port != 25:
            mail_errors.append("incorrect port")
        if (hasattr(mail_host, "smtp_userid") and mail_host.smtp_userid) or (
                hasattr(mail_host, "smtp_uid") and mail_host.smtp_uid):
            mail_errors.append("user entered")
        if (hasattr(mail_host, "smtp_pass") and mail_host.smtp_pass) or (
                hasattr(mail_host, "smtp_pwd") and mail_host.smtp_pwd):
            mail_errors.append("password entered")
        if not plone_site.email_from_address or plone_site.email_from_address == "postmaster@localhost":
            mail_errors.append("bad mail")
        connection.write(str(len(mail_errors)))
        app._p_jar.close()


def check_upgrade_steps(connection, plone_path=None):
    """Check if all upgrade steps are ran."""
    app, plone_site = get_plone_site(connection, plone_path)
    not_upgraded = 0
    if plone_site:
        setSite(plone_site)
        ps = plone_site.portal_setup
        # qit = plone_site.portal_quickinstaller
        for profile_id in ps.listProfilesWithUpgrades():
            is_installed = ps.getLastVersionForProfile(profile_id) != 'unknown'
            if is_installed:
                # latest_upgrade_step = qit.getLatestUpgradeStep(profile_id)
                # if latest_upgrade_step != 'unknown':
                upgrades = ps.listUpgrades(profile_id)
                if upgrades:
                    not_upgraded += 1

        connection.write(str(not_upgraded))
    app._p_jar.close()


def creation_date_plonesite(dconnection, plone_path=None):
    """Get creation date of plonesite object"""
    app, plone_site = get_plone_site(dconnection, plone_path)
    if plone_site:
        setSite(plone_site)
        creation_date = plone_site.creation_date
        # creation_date.Date() if you want a yyyy/mm/dd format
        dconnection.write(str(creation_date))
    app._p_jar.close()


def last_login_time(dconnection, plone_path=None):
    """Get last login time user"""
    app, plone_site = get_plone_site(dconnection, plone_path)
    if plone_site:
        setSite(plone_site)
        users = get_users(plone_site)
        last_login = ""
        for user in users:
            if not last_login:
                last_login = user.getProperty("last_login_time")
            if user.getProperty("last_login_time") > last_login:
                last_login = user.getProperty("last_login_time")
        dconnection.write(str(last_login))
    app._p_jar.close()


def last_modified_plone_object_time(dconnection, plone_path=None):
    """Get last modified plone object time"""
    app, plone_site = get_plone_site(dconnection, plone_path)
    if plone_site:
        setSite(plone_site)
        pc = plone_site.portal_catalog
        query = {}
        query['path'] = '/'.join(plone_site.getPhysicalPath())
        query['sort_on'] = 'modified'
        query['sort_order'] = 'reverse'
        query['sort_limit'] = 1
        brain = pc(query)[0]
        dconnection.write(str(brain.modified))
    app._p_jar.close()


def last_modified_zope_object_time(dconnection, plone_path=None):
    """Get last modified zope object time"""
    from zope.globalrequest import setRequest
    from Testing import makerequest
    app = App()
    app = makerequest.makerequest(app)
    # support plone.subrequest
    app.REQUEST['PARENTS'] = [app]
    setRequest(app.REQUEST)
    container = app.unrestrictedTraverse('/')
    undoable_transactions = container.undoable_transactions()
    if len(undoable_transactions) > 1:
        last_modified = undoable_transactions[0]['time']
        dconnection.write(str(last_modified))
    app._p_jar.close()


def dates(dconnection, plone_path=None):
    """Return all date probes"""
    date_probes = [
        'creation_date_plonesite',
        'last_login_time',
        'last_modified_plone_object_time',
        'last_modified_zope_object_time',
    ]
    for name, probe in getUtilitiesFor(IZ3MonitorPlugin):
        if name not in date_probes:
            continue
        argspec = inspect.getargspec(probe)
        tempStream = StringIO()
        probe(tempStream)
        beautify_return_values(dconnection, tempStream, name)
